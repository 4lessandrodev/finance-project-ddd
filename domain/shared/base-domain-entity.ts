export abstract class BaseDomainEntity {
  constructor(
    public createdAt?: Date,
    public updatedAt?: Date,
    public isDeleted?: boolean,
    public deletedAt?: Date | null,
  ) {
    this.createdAt = createdAt ?? new Date();
    this.updatedAt = updatedAt ?? new Date();
    this.isDeleted = isDeleted ?? false;
    this.deletedAt = isDeleted ? deletedAt : null;
  }
}
