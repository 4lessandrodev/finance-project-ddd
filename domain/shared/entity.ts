import { BaseDomainEntity, UniqueEntityID } from './index';

const isEntity = (v: any): v is Entity<any> => {
  return v instanceof Entity;
};

export abstract class Entity<T extends BaseDomainEntity> {
  protected readonly _id: UniqueEntityID;
  protected readonly props: T;

  constructor(props: T, id?: UniqueEntityID) {
    this._id = id ? id : new UniqueEntityID();
    this.props = props;
  }

  get createdAt(): Date {
    return this.props.createdAt ?? new Date();
  }

  get updatedAt(): Date {
    return this.props.updatedAt ?? new Date();
  }

  get isDeleted(): boolean {
    return this.props.isDeleted ?? false;
  }

  public equals(object?: Entity<T>): boolean {
    if (object == null || object == undefined) {
      return false;
    }

    if (this === object) {
      return true;
    }

    if (!isEntity(object)) {
      return false;
    }

    return this._id.equals(object._id);
  }
}
