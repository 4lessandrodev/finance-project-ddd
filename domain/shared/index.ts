export * from './aggregate-root';
export * from './entity';
export * from './identifier';
export * from './result';
export * from './unique-entity-id';
export * from './events/domain-events';
export * from './events/IDomainEvent';
export * from './events/ihandle';
export * from './value-object';
export * from './base-domain-entity';
export * from './iuse-case';
