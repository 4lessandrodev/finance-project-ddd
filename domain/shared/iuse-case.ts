export interface IUseCase<Request, Response> {
  execute: (request: Request) => Promise<Response>;
}
