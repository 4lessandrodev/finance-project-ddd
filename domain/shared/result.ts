export class Result<T> {
  public isSuccess: boolean;
  public isFailure: boolean;
  public error: T | string;
  private _value: T;

  public constructor(isSuccess: boolean, error?: T | string | null, value?: T) {
    if (isSuccess && error) {
      throw new Error(
        'InvalidOperation: A result cannot be successful and contain an error',
      );
    }
    if (!isSuccess && !error) {
      throw new Error(
        'InvalidOperation: A failing result needs to contain an error message',
      );
    }

    this.isSuccess = isSuccess;
    this.isFailure = !isSuccess;
    this.error = error as T;
    this._value = value as T;

    Object.freeze(this);
  }
  /**
   * If success returns an instance of provided class.
   */
  public getResult(): T {
    if (!this.isSuccess) {
      console.log(this.error);
      throw new Error(
        'Can not get the value of an error result. Use errorValue instead.',
      );
    }

    return this._value;
  }

  public errorValue(): T {
    return this.error as T;
  }

  public static ok<U>(value?: U): Result<U> {
    return new Result<U>(true, null, value);
  }

  public static fail<U>(error: string): Result<U> {
    return new Result<U>(false, error);
  }

  public static combine(results: Result<any>[]): Result<any> {
    for (const result of results) {
      if (result.isFailure) return result;
    }
    return Result.ok();
  }
}
