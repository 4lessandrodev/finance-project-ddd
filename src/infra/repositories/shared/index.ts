export * from './core/interfaces/generic-repository.interface';
