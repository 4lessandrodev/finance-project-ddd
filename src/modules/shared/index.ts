export * from './utils/error-messages/messages';
export * from './interfaces/budget-box-model.interface';
export * from './interfaces/entity-mock.interface';
export * from './interfaces/reason-model.interface';
export * from './interfaces/transaction-model.interface';
export * from './interfaces/user-model.interface';
export * from './interfaces/domain-service.interface';
export * from './domain/budget-box-connection.interface';
export * from './infra/connections/budget-box-connection';
export * from './infra/shared.module';
export * from './proxies/base.proxy';
export * from './infra/connections/connection';