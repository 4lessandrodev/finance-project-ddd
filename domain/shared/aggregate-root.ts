import { Entity } from './entity';
import { DomainEvents } from './events/domain-events';
import { IDomainEvent } from './events/IDomainEvent';
import { UniqueEntityID } from './unique-entity-id';

export abstract class AggregateRoot<T> extends Entity<T> {
  private _domainEvents: IDomainEvent[] = [];

  get id(): UniqueEntityID {
    return this._id;
  }

  get domainEvents(): IDomainEvent[] {
    return this._domainEvents;
  }

  protected addDomainEvent(domainEvent: IDomainEvent): void {
    // Add the domain event to this aggregate's list of domain events
    this._domainEvents.push(domainEvent);
    // Add this aggregate instance to the domain event's list of aggregates who's
    // events it eventually needs to dispatch.
    DomainEvents.markAggregateForDispatch(this);
    // Log the domain event
    this.logDomainEventAdded(domainEvent);
  }

  public clearEvents(): void {
    this._domainEvents.splice(0, this._domainEvents.length);
  }

  private logDomainEventAdded(domainEvent: IDomainEvent): void {
    const thisClass = Reflect.getPrototypeOf(this);
    const domainEventClass = Reflect.getPrototypeOf(domainEvent);
    console.info(
      '[Domain Event Created]:',
      thisClass?.constructor.name,
      '==>',
      domainEventClass?.constructor.name,
    );
  }
}
