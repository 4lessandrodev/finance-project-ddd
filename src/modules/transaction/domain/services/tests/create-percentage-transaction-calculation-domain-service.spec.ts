import { IBudgetBox, IBudgetBoxConnection } from "@modules/shared";
import budgetBoxConnectionMock from "@modules/shared/domain/tests/mocks/budget-box-connection.mock";
import CreatePercentageTransactionCalculationDomainService from "../create-percentage-transaction-calculation.domain-service";

describe('create-percentage-transaction-calculation.domain-service', () => {

	let fakeConnection: IBudgetBoxConnection;

	beforeEach(() => {
		fakeConnection = budgetBoxConnectionMock;
	});

	it('should calculate percentage for each budget box', async () => {

		const data = [10, 10, 20, 20, 40].map((val) => ({
			id: `valid_id-${val}`,
			isPercentage: true,
			description: `valid_name-${val}`,
			budgetPercentage: val
		}));

		const invalidModel = {
			id: 'invalid_id',
			isPercentage: false,
			description: 'valid_name',
			budgetPercentage: 100
		};

		jest.spyOn(fakeConnection, 'findBudgetBoxesByUserId').mockResolvedValueOnce(
			[...data, invalidModel] as IBudgetBox[]
		);

		const domainService = new CreatePercentageTransactionCalculationDomainService(fakeConnection);

		const result = await domainService.execute({
			total: 100,
			userId: 'valid_id'
		});

		const calculation = result.map((calc) => calc.toObject());
		
		const total = result.reduce((total, calculation) => total + calculation.currency.value, 0);
		expect(total).toBe(100);
		expect(calculation).toHaveLength(5);
		expect(calculation).toMatchSnapshot();
	});
});
