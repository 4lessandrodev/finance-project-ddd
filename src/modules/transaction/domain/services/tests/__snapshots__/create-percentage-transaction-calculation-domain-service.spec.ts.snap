// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-percentage-transaction-calculation.domain-service should calculate percentage for each budget box 1`] = `
Array [
  Object {
    "budgetBoxId": "valid_id-10",
    "budgetBoxName": "valid_name-10",
    "currency": Object {
      "currency": "BRL",
      "value": 10,
    },
  },
  Object {
    "budgetBoxId": "valid_id-10",
    "budgetBoxName": "valid_name-10",
    "currency": Object {
      "currency": "BRL",
      "value": 10,
    },
  },
  Object {
    "budgetBoxId": "valid_id-20",
    "budgetBoxName": "valid_name-20",
    "currency": Object {
      "currency": "BRL",
      "value": 20,
    },
  },
  Object {
    "budgetBoxId": "valid_id-20",
    "budgetBoxName": "valid_name-20",
    "currency": Object {
      "currency": "BRL",
      "value": 20,
    },
  },
  Object {
    "budgetBoxId": "valid_id-40",
    "budgetBoxName": "valid_name-40",
    "currency": Object {
      "currency": "BRL",
      "value": 40,
    },
  },
]
`;
